### PR TITLE
Add reference providers doc

### LINKED ISSUE #9993: Smart picker integration
## Use case 1

As a Nextcloud app developer I want to integrate my app's capabilities into the smart picker feature. How do I do that?

## Use case 2

As a Nextcloud app developer I want the smart picker to show inside my app's UI. How do I do that?

cc @DaphneMuller

### PR DESCRIPTION
**Pull request by julien-nc**
closes #9993

Add a new page about:

* [x] display link previews in apps and clients
* [x] display the smart picker in apps and clients
* [x] implement reference providers to
    * [x] extend link previews support
    * [x] extend the smart picker

### CODE REVIEW

File: developer_manual/digging_deeper/reference.rst [hunk 1]
```diff
@@ -0,0 +1,527 @@
...
+
+.. code-block:: bash
+
+    curl -u USER:PASSWD -H "Accept: application/json" -H "ocs-apirequest: true" \
+        "https://my.nextcloud.org/ocs/v2.php/references/resolve?reference=https://github.com"
+
+will return an OCS response with that data:
+
+.. code-block:: json
+
```
Comment (ChristophWurst) [inline developer_manual/digging_deeper/reference.rst:133]: missing an opening { to become valid json

Comment (julien-nc) [reply developer_manual/digging_deeper/reference.rst:133]: It is just part of the json object. But yeah, let's mention the entire OCS response so it's valid json. :+1:

File: developer_manual/digging_deeper/reference.rst [hunk 2]
```diff
@@ -0,0 +1,604 @@
...
+* the ones implementing their own custom picker component
+
+In summary, reference providers can be registered by apps to
+
+* add support for new kinds of HTTP links
+    * resolve the links, get information on the link targets
+    * optionally provide their own reference widgets to have a custom preview rendering
+* extend the Smart Picker
+    * use existing Unified Search providers
+    * or optionally register custom picker components to have a specific user interface
```
Comment (ChristophWurst) [inline developer_manual/digging_deeper/reference.rst:33]: does this nested list render as expected? I think there has to be a blank line between the first level items

File: developer_manual/digging_deeper/reference.rst [hunk 3]
```diff
@@ -0,0 +1,604 @@
...
+you need to make sure the ``OCP\Collaboration\Reference\RenderReferenceEvent`` is dispatched
+before you load the page where you want to display link previews.
+
+For example, you can place this before returning your TemplateResponse in your controller:
+
+.. code-block:: php
+
+    $this->eventDispatcher->dispatchTyped(new OCP\Collaboration\Reference\RenderReferenceEvent());
+
+This is done in Text and Talk if you need more examples.
```
Comment (ChristophWurst) [inline developer_manual/digging_deeper/reference.rst:55]: link to the code?

File: developer_manual/digging_deeper/reference.rst [hunk 4]
```diff
@@ -0,0 +1,604 @@
...
+            "description": "30 mai 2012 - A team of explorers discover a clue to the origins of mankind on Earth, leading them on a journey to the darkest corners of the universe. There, they must fight a terrifying battle to save the future of the human race.",
+            "thumb": "https://my.nextcloud.org/apps/integration_tmdb/t/p/w500/qsYQflQhOuhDpQ0W2aOcwqgDAeI.jpg?fallbackName=???",
+            "link": "https://www.themoviedb.org/movie/70981"
+          },
+          "accessible": true
+        }
+      }
+    }
+
+Use the Smart Picker in your app
```
Comment (ChristophWurst) [inline developer_manual/digging_deeper/reference.rst:223]: the code block is still active here. add at least one more blank linkes above this line

Comment (julien-nc) [reply developer_manual/digging_deeper/reference.rst:223]: It was rendered correctly, still added a blank line, why not.

File: developer_manual/digging_deeper/reference.rst (hunk 3 above)
Comment (julien-nc) [reply developer_manual/digging_deeper/reference.rst:55]: :+1: Done

File: developer_manual/digging_deeper/reference.rst (hunk 2 above)
Comment (julien-nc) [reply developer_manual/digging_deeper/reference.rst:33]: image

It's fine without blank lines.

Comment (ChristophWurst) [reply developer_manual/digging_deeper/reference.rst:33]: okay, I'm sorry. for the nested list I wrote the other day sphinx wouldn't render correctly without the line :shrug:

**Review by ChristophWurst — APPROVED** [review]

**Pull request merged**